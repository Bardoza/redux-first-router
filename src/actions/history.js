// @flow
import { CALL_HISTORY } from '../types'

export const push = (path: string, state: ?Object) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'push',
    args: [path, state]
  }
})

export const replace = (path: string, state: ?Object) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'replace',
    args: [path, state]
  }
})

export const jump = (n: number | string, state: ?Object, byIndex: ?boolean, kind: ?string) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'jump',
    args: [n, state, byIndex, kind]
  }
})

export const reset = (entries: Array<string | Object>, index: ?number, kind: ?string) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'reset',
    args: [entries, index, kind]
  }
})

export const setState = (state: Object | Function, n: ?(number | string), byIndex: ?boolean) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'setState',
    args: [state, n, byIndex]
  }
})

export const back = (state: ?(Object | Function)) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'back',
    args: [state]
  }
})

export const next = (state: ?(Object | Function)) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'next',
    args: [state]
  }
})


// NOTE: see `src/utils/formatRoutes.js` for implemenation of corresponding pathlessRoute
