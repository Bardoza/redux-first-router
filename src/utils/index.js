export { default as isHydrate } from './isHydrate'
export { default as isAction } from './isAction'
export { default as isNotFound } from './isNotFound'
export { default as isServer } from './isServer'
export { default as isRedirect } from './isRedirect'

export { default as actionToUrl } from './actionToUrl'
export { default as urlToAction } from './urlToAction'

export { default as doesRedirect } from './doesRedirect'
export { default as shouldTransition } from './shouldTransition'

export { default as matchUrl } from './matchUrl'
export { default as compileUrl } from './compileUrl'

export { default as formatRoutes, formatRoute } from './formatRoutes'
export { default as typeToScene } from './typeToScene'

export { default as noOp } from './noOp'

export { default as createSelector } from './createSelector'
export { default as createAction } from './createAction'

export { default as redirectShortcut } from './redirectShortcut'
export { default as callRoute } from './callRoute'
