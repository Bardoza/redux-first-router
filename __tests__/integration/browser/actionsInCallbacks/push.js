import createTest, { resetBrowser } from '../../../../__helpers__/createTest'
import { push } from '../../../../src/actions'

beforeEach(resetBrowser)

createTest('push before enter', {
  FIRST: '/',
  SECOND: {
    path: '/second',
    beforeEnter: () => push('/redirected'),
    thunk: function() {}
  }
}, { testBrowser: true })

createTest('push after enter', {
  FIRST: '/',
  SECOND: {
    path: '/second',
    thunk: () => push('/redirected'),
    onComplete: function() {}
  }
}, { testBrowser: true })

createTest('push before enter (on firstRoute)', {
  FIRST: {
    path: '/',
    beforeEnter: () => push('/redirected'),
    thunk: function() {}
  }
}, { testBrowser: true })

createTest('push after enter (on firstRoute)', {
  FIRST: {
    path: '/',
    thunk: () => push('/redirected'),
    onComplete: function() {}
  }
}, { testBrowser: true })
