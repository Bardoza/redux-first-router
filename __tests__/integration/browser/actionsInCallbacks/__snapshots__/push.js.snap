// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`push after enter (on firstRoute): 1 - routes - FIRST - onComplete 1`] = `0`;

exports[`push after enter (on firstRoute): 1 - routes - FIRST - thunk 1`] = `1`;

exports[`push after enter (on firstRoute): 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`push after enter (on firstRoute): 1 - routes - REDIRECTED - onComplete 1`] = `1`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/redirected",
      "scene": "",
      "search": "",
      "url": "/redirected",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "REDIRECTED",
  },
]
`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - history_index 1`] = `1`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - initial_state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "init",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "@@redux/INIT",
}
`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - response 1`] = `
Object {
  "payload": "redirect_complete",
  "type": "REDIRECTED_COMPLETE",
}
`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/redirected",
      Object {},
      "345678",
    ],
  ],
  "index": 1,
}
`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/redirected",
          "scene": "",
          "search": "",
          "url": "/redirected",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "REDIRECTED",
      },
    ],
    "from": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "hash": "",
    "index": 1,
    "key": "345678",
    "kind": "load",
    "length": 2,
    "n": 1,
    "params": Object {},
    "pathname": "/redirected",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "ready": true,
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 302,
    "type": "REDIRECTED",
    "universal": false,
    "url": "/redirected",
  },
  "title": "REDIRECTED_COMPLETE - \\"redirect_complete\\"",
}
`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - title 1`] = `"REDIRECTED"`;

exports[`push after enter (on firstRoute): firstRoute - /first - 1 - windowLocation 1`] = `"/redirected"`;

exports[`push after enter: 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/second",
      "scene": "",
      "search": "",
      "url": "/second",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "SECOND",
  },
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/redirected",
      "scene": "",
      "search": "",
      "url": "/redirected",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "REDIRECTED",
  },
]
`;

exports[`push after enter: 1 - history_index 1`] = `2`;

exports[`push after enter: 1 - response 1`] = `
Object {
  "payload": "redirect_complete",
  "type": "REDIRECTED_COMPLETE",
}
`;

exports[`push after enter: 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`push after enter: 1 - routes - REDIRECTED - onComplete 1`] = `1`;

exports[`push after enter: 1 - routes - SECOND - onComplete 1`] = `0`;

exports[`push after enter: 1 - routes - SECOND - thunk 1`] = `1`;

exports[`push after enter: 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/redirected",
      Object {},
      "345678",
    ],
  ],
  "index": 2,
}
`;

exports[`push after enter: 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/redirected",
          "scene": "",
          "search": "",
          "url": "/redirected",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "REDIRECTED",
      },
    ],
    "from": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "hash": "",
    "index": 2,
    "key": "345678",
    "kind": "replace",
    "length": 3,
    "n": 1,
    "params": Object {},
    "pathname": "/redirected",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "query": Object {},
    "ready": true,
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 302,
    "type": "REDIRECTED",
    "universal": false,
    "url": "/redirected",
  },
  "title": "REDIRECTED_COMPLETE - \\"redirect_complete\\"",
}
`;

exports[`push after enter: 1 - title 1`] = `"REDIRECTED"`;

exports[`push after enter: 1 - windowLocation 1`] = `"/redirected"`;

exports[`push after enter: firstRoute - /first - 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
]
`;

exports[`push after enter: firstRoute - /first - 1 - history_index 1`] = `0`;

exports[`push after enter: firstRoute - /first - 1 - initial_state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "init",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "@@redux/INIT",
}
`;

exports[`push after enter: firstRoute - /first - 1 - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "index": 0,
    "key": "345678",
    "kind": "load",
    "length": 1,
    "n": 1,
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "FIRST",
}
`;

exports[`push after enter: firstRoute - /first - 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
  ],
  "index": 0,
}
`;

exports[`push after enter: firstRoute - /first - 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "load",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`push after enter: firstRoute - /first - 1 - title 1`] = `"FIRST"`;

exports[`push after enter: firstRoute - /first - 1 - windowLocation 1`] = `"/"`;

exports[`push before enter (on firstRoute): 1 - routes - FIRST - beforeEnter 1`] = `1`;

exports[`push before enter (on firstRoute): 1 - routes - FIRST - thunk 1`] = `0`;

exports[`push before enter (on firstRoute): 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`push before enter (on firstRoute): 1 - routes - REDIRECTED - onComplete 1`] = `1`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/redirected",
      "scene": "",
      "search": "",
      "url": "/redirected",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "REDIRECTED",
  },
]
`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - history_index 1`] = `1`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - initial_state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "init",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "@@redux/INIT",
}
`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - response 1`] = `
Object {
  "payload": "redirect_complete",
  "type": "REDIRECTED_COMPLETE",
}
`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/redirected",
      Object {},
      "345678",
    ],
  ],
  "index": 1,
}
`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/redirected",
          "scene": "",
          "search": "",
          "url": "/redirected",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "REDIRECTED",
      },
    ],
    "from": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "hash": "",
    "index": 1,
    "key": "345678",
    "kind": "load",
    "length": 2,
    "n": 1,
    "params": Object {},
    "pathname": "/redirected",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "ready": true,
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 302,
    "type": "REDIRECTED",
    "universal": false,
    "url": "/redirected",
  },
  "title": "REDIRECTED_COMPLETE - \\"redirect_complete\\"",
}
`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - title 1`] = `"REDIRECTED"`;

exports[`push before enter (on firstRoute): firstRoute - /first - 1 - windowLocation 1`] = `"/redirected"`;

exports[`push before enter: 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/redirected",
      "scene": "",
      "search": "",
      "url": "/redirected",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "REDIRECTED",
  },
]
`;

exports[`push before enter: 1 - history_index 1`] = `1`;

exports[`push before enter: 1 - response 1`] = `
Object {
  "payload": "redirect_complete",
  "type": "REDIRECTED_COMPLETE",
}
`;

exports[`push before enter: 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`push before enter: 1 - routes - REDIRECTED - onComplete 1`] = `1`;

exports[`push before enter: 1 - routes - SECOND - beforeEnter 1`] = `1`;

exports[`push before enter: 1 - routes - SECOND - thunk 1`] = `0`;

exports[`push before enter: 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/redirected",
      Object {},
      "345678",
    ],
  ],
  "index": 1,
}
`;

exports[`push before enter: 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/redirected",
          "scene": "",
          "search": "",
          "url": "/redirected",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "REDIRECTED",
      },
    ],
    "from": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "hash": "",
    "index": 1,
    "key": "345678",
    "kind": "replace",
    "length": 2,
    "n": 1,
    "params": Object {},
    "pathname": "/redirected",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "query": Object {},
    "ready": true,
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 302,
    "type": "REDIRECTED",
    "universal": false,
    "url": "/redirected",
  },
  "title": "REDIRECTED_COMPLETE - \\"redirect_complete\\"",
}
`;

exports[`push before enter: 1 - title 1`] = `"REDIRECTED"`;

exports[`push before enter: 1 - windowLocation 1`] = `"/redirected"`;

exports[`push before enter: firstRoute - /first - 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
]
`;

exports[`push before enter: firstRoute - /first - 1 - history_index 1`] = `0`;

exports[`push before enter: firstRoute - /first - 1 - initial_state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "init",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "@@redux/INIT",
}
`;

exports[`push before enter: firstRoute - /first - 1 - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "index": 0,
    "key": "345678",
    "kind": "load",
    "length": 1,
    "n": 1,
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "FIRST",
}
`;

exports[`push before enter: firstRoute - /first - 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
  ],
  "index": 0,
}
`;

exports[`push before enter: firstRoute - /first - 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "load",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`push before enter: firstRoute - /first - 1 - title 1`] = `"FIRST"`;

exports[`push before enter: firstRoute - /first - 1 - windowLocation 1`] = `"/"`;
